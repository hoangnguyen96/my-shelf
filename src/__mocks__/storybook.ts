export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};
