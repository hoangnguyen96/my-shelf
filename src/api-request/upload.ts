import { MESSAGES } from "@app/constants";

export const generateImageUpload = async (file: File) => {
  try {
    const formData = new FormData();
    formData.append("image", file);

    const response = await fetch(
      `https://api.imgbb.com/1/upload?key=${process.env.NEXT_PUBLIC_IMGBB_API_KEY}`,
      {
        method: "POST",
        body: formData,
      }
    );

    if (!response.ok) {
      throw new Error(MESSAGES.UPLOAD_FAILED);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    throw new Error(MESSAGES.ERROR_UPLOAD);
  }
};
