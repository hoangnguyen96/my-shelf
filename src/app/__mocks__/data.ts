export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
