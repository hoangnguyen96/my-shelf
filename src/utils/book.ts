import { DEFAULT_LIMIT, TYPE_SEARCH } from "@app/constants";
import { BookType } from "@app/interface";

export const filterBooksOnShelf = (
  allBooks: BookType[],
  shelfBooks: string[]
) => allBooks.filter((item) => shelfBooks.includes(item.id));

export const filterBooksOnShelfByParams = (
  booksOnShelf: BookType[],
  type: string,
  value: string
) =>
  booksOnShelf.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const filterBooksFavorite = (
  allBooks: BookType[],
  favorites: string[]
) => allBooks.filter((item) => favorites.includes(item.id));

export const filterBooksFavoriteByParams = (
  booksFavorites: BookType[],
  type: string,
  value: string
) =>
  booksFavorites.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const dividePaginationBooks = (list: BookType[], limit?: number) => {
  return list.reduce((acc: BookType[][], _, i, self) => {
    if (!(i % (limit || DEFAULT_LIMIT))) {
      return [...acc, self.slice(i, i + (limit || DEFAULT_LIMIT))];
    }
    return acc;
  }, []);
};
