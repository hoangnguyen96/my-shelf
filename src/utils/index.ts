export * from "./book";
export * from "./date";
export * from "./error";
export * from "./search";
export * from "./user";
export * from "./validators";
