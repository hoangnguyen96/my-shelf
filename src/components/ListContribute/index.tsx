"use client";

import { getAllBook, getUserById } from "@app/api-request";
import { Cart, LoadingIndicator } from "@app/components/common";
import { BookType, User } from "@app/models";
import { getThreeTopBook } from "@app/utils";
import { Flex } from "@chakra-ui/react";
import { memo, useEffect, useState } from "react";
import { useSession } from "next-auth/react";

const ListContribute = () => {
  const [dataTop, setDataTop] = useState<BookType[] | null>(null);
  const [dataUserById, setDataUserById] = useState<User | null>(null);

  const { data: session } = useSession();

  const fetchData = async () => {
    try {
      const dataUserById = (await getUserById(session?.user?.id as string)) as User;
      const dataBooks = await getAllBook();
      const dataTop = getThreeTopBook(dataBooks);

      setDataUserById(dataUserById);
      setDataTop(dataTop);
    } catch (error) {
      console.error("Failed to fetch data:", error);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  if (!dataTop || !dataUserById) {
    return <LoadingIndicator />;
  }

  return (
    <Flex gap="39px" justifyContent="space-between">
      {dataTop.map((item: BookType) => {
        const { id, title, author, imageUrl, publicationYear, rating } = item;

        return (
          <Cart
            key={id}
            id={id}
            title={title}
            author={author}
            imageUrl={imageUrl}
            publicationYear={publicationYear}
            rating={rating}
            isFavorite={dataUserById?.favorites?.includes(id)}
            isContribute={true}
          />
        );
      })}
    </Flex>
  );
};

export default memo(ListContribute);
