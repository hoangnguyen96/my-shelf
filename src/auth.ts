import NextAuth from "next-auth";
import { authConfig } from "./auth.config";
import Credentials from "next-auth/providers/credentials";
import { z } from "zod";
import bcrypt from "bcryptjs";
import { MESSAGES } from "./constants";
import { getUserByEmail } from "@app/features/dashboard/actions";
import { User } from "./models";

export const { handlers, auth, signIn, signOut } = NextAuth({
  ...authConfig,
  secret: process.env.AUTH_SECRET,
  providers: [
    Credentials({
      name: "Credentials",
      credentials: {
        email: { label: "Email", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        if (!credentials) {
          return null;
        }

        const parsedCredentials = z
          .object({ email: z.string().email(), password: z.string().min(6) })
          .safeParse(credentials);

        if (parsedCredentials.success) {
          const { email, password } = parsedCredentials.data;
          const user = (await getUserByEmail(email)) as User;

          if (!user || !user.password) {
            throw new Error(MESSAGES.LOGIN_NOTFOUND);
          }

          const passwordsMatch = await bcrypt.compare(password, user.password);

          if (passwordsMatch)
            return {
              id: user.userId,
              name: user.username,
              isAdmin: user.isAdmin,
              image: user.avatar,
              email: user.email,
            };
        }

        throw new Error(MESSAGES.INVALID_CREDENTIALS);
      },
    }),
  ],
});
