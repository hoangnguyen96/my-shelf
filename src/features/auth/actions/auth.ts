"use server";

import { AuthError } from "next-auth";
import { MESSAGES } from "@app/constants";
import { User } from "@app/interface";
import { signIn, signOut } from "@app/auth";

export const authenticate = async (formData: Partial<User>) => {
  try {
    await signIn("credentials", formData);
  } catch (error) {
    if (error instanceof AuthError) {
      return MESSAGES.EMAIL_PASSWORD_INVALID;
    }
    throw error;
  }
};

export const logout = async () => {
  await signOut();
};
