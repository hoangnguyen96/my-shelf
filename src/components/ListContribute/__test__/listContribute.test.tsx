import { auth } from "@app/auth";
import { act, render } from "@testing-library/react";
import { getAllBook, getUserById } from "@app/api-request";
import * as utils from "@app/utils";
import { DATA_BOOKS, DATA_USER } from "@app/__mocks__/data";
import ListContribute from "..";
import { useSession } from "next-auth/react";

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("@app/api-request", () => ({
  getAllBook: jest.fn(),
  getUserById: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  getThreeTopBook: jest.fn(),
}));

jest.mock("@app/actions/auth", () => ({
  logout: jest.fn(),
}));

describe("Contribute Three Top Book", () => {
  const mockThreeTopBook = DATA_BOOKS.sort((a, b) => {
    return (
      new Date(b.createdDate).getTime() - new Date(a.createdDate).getTime()
    );
  }).slice(0, 3);

  (useSession as jest.Mock).mockReturnValue({
    data: {
      user: {
        isAdmin: true,
        email: "admin@gmail.com",
        id: "3733403",
        name: "admin",
        image: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
      expires: "2024-12-31T23:59:59.999Z",
    },
    status: "authenticated",
  });

  beforeEach(() => {
    jest.clearAllMocks();
    (getUserById as jest.Mock).mockResolvedValue(DATA_USER[0]);

    (getAllBook as jest.Mock).mockResolvedValue(DATA_BOOKS);
    jest.spyOn(utils, "getThreeTopBook").mockReturnValue(mockThreeTopBook);
  });

  it("Should render correctly snapshot", async () => {
    await act(async () => {
      const { container } = render(<ListContribute />);
      expect(container).toMatchSnapshot();
    });
  });
});
