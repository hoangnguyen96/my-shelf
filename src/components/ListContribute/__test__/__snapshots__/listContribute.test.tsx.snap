// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contribute Three Top Book Should render correctly snapshot 1`] = `<div />`;

exports[`Contribute Three Top Book Should render correctly snapshot when error 1`] = `<div />`;
